packages/headless/src/controllers/controller/headless-controller.ts

```diff
@@ -15,13 +15,16 @@ export interface Controller {
 export function buildController<T extends object>(
   engine: CoreEngine<T>
 ): Controller {
-  let prevState: string;
+  const prevState: Map<Symbol, string> = new Map();
 
-  const hasStateChanged = (currentState: Record<string, unknown>): boolean => {
+  const hasStateChanged = (
+    currentState: Record<string, unknown>,
+    symbol: Symbol
+  ): boolean => {
     try {
```

Comment:
I know it's too late since this was in a different PR, but I think we need to discuss this.

Rather than subscribing to the engine inside the `subscribe` method and rather than saving multiple copies of the state for each `subscribe` call, Ithink we should keep a map of listeners and called them all at once.

E.g.:
```ts
export function buildController<...>(...): Controller {
  const listeners: Map<Symbol, () => void> = new Map();
  let prevState: string;
  let unsubscribe: () => void | null = null;

  return {
    subscribe(listener: () => void) {
      const symbol = new Symbol();
      listeners.set(symbol, listener);
      if (listeners.size === 1) {
        unsubscribe = engine.subscribe(() => {
          // compare state and call all listeners if it changed
        });
      }
      return () => {
        listeners.delete(symbol);
        if (listeners.size === 0) {
          unsubscribe();
        }
      }
    }
  };
}
```